File: 2024/JS/Day1/solutionB.js
const fs = require("fs");

fs.readFile("../../data/day1.txt", (err, data) => {
  const unorderedLeft = [];
  const unorderedRight = [];
  data
    .toString()
    .split("\n")
    .slice(0, -1)
    .forEach((row) => {
      const [l, r] = row.split("   ");
      unorderedLeft.push(l);
      unorderedRight.push(r);
    });

  const orderedLeft = unorderedLeft.sort((a, b) => a - b);
  const orderedRight = unorderedRight.sort((a, b) => a - b);

  const ref = {};
  const multiples = orderedLeft.map((l, i) => {
    if (ref[l]) return l * ref[l];

    ref[l] = orderedRight.filter((r) => r === l).length;
    return l * ref[l];
  });

  const sum = multiples.reduce((acc, curr) => acc + curr, 0);

  console.log(sum);
});

File: 2024/Python/Day1/solutionB.py
left = []
right = []

with open("../../data/day1.txt") as file:
    for line in file:
        [l, r] = line.rstrip('\n').split("   ")
        left.append(int(l))
        right.append(int(r))

left.sort()
right.sort()

ref = {}
multiples = []

for l in left:
    if l in ref:
        multiples.append(l * ref[l])
    else:
        ref[l] = right.count(l)
        multiples.append(l * ref[l])

print(sum(multiples))
